src/middlewared/middlewared/plugins/account_/privilege.py

```diff
@@ -0,0 +1,161 @@
+import enum
+import errno
+
+from middlewared.schema import accepts, Bool, Dict, Int, List, Ref, Str, Patch
+from middlewared.service import CallError, CRUDService, private, ValidationErrors
+import middlewared.sqlalchemy as sa
+
+
+class BuiltinPrivileges(enum.Enum):
+    LOCAL_ADMINISTRATOR = "LOCAL_ADMINISTRATOR"
+
+
+class PrivilegeModel(sa.Model):
+    __tablename__ = "account_privilege"
+
+    id = sa.Column(sa.Integer(), primary_key=True)
+    builtin_name = sa.Column(sa.String(200), nullable=True)
+    name = sa.Column(sa.String(200))
+    local_groups = sa.Column(sa.JSON(type=list))
+    ds_groups = sa.Column(sa.JSON(type=list))
+    allowlist = sa.Column(sa.JSON(type=list))
+    web_shell = sa.Column(sa.Boolean())
+
+
+class PrivilegeService(CRUDService):
+
+    keys = {}
+
+    class Config:
+        namespace = "privilege"
+        datastore = "account.privilege"
+        datastore_extend = "privilege.item_extend"
+        datastore_extend_context = "privilege.item_extend_context"
+        cli_namespace = "auth.privilege"
+
+    ENTRY = Dict(
+        "privilege_entry",
+        Int("id"),
+        Str("builtin_name", null=True),
+        Str("name", required=True, empty=False),
+        List("local_groups", items=[Int("local_group")]),
+        List("ds_groups", items=[Int("ds_group")]),
+        List("allowlist", items=[Ref("allowlist_item")]),
+        Bool("web_shell", required=True),
+    )
+
+    @private
+    async def item_extend_context(self, rows, extra):
+        return {
+            "local_groups": {
+                g["id"]: {"id": g["id"], "name": g["group"]}
+                for g in await self.middleware.call("datastore.query", "account.bsdgroups", [], {"prefix": "bsdgrp_"})
+            },
```

Comment:
Why aren't you using group.query? You can get all local and DS groups at once there.

```
midclt call group.query '[]' '{"extra": {"additional_information": ["DS"]}}'
```
The `local` key for each entry differentiates between local and directory services groups:

```
  {
    "id": 90101103,
    "gid": 100001103,
    "name": "BILLY\\dnsupdateproxy",
    "group": "BILLY\\dnsupdateproxy",
    "builtin": false,
    "sudo": false,
    "sudo_nopasswd": false,
    "sudo_commands": [],
    "users": [],
    "local": false,
    "id_type_both": true,
    "nt_name": null,
    "sid": null
  }
  ```

Reply:
There is a chance that a particular GID may not be resolvable through the query results (because we're hitting a cache for directory services case). In this situation you need to perform a second query for the group specifically by gid `[["gid", '=", gid]]`. 

There is special behavior in this case which will go directly do the getgrgid request, convert into our normal group dict, insert entry in cache, and return entry.

Reply:
@anodos325 is `midclt call group.query '[]' '{"extra": {"additional_information": ["DS"]}}'` fast enough, does it always come from the cache?

Reply:
Yes, it's always from cache. The only time it doesn't hit cache is if you have a filter like above.